```{index} Operators
```

# Operators


```{contents}
:local:
:depth: 2
```


An **operator** applies a rule to operands, the values it works on. Operand types affect the rule: `12 / 5` performs integer division, while `12m / 5` performs decimal division. An invoice calculation must use both the right quantities and the intended arithmetic.


```{index} arithmetic operators; precedence
```

(arithmetic-operators)=

## Arithmetic and Grouping



| Operator | Operation | Example | Result |
| --- | --- | --- | --- |
| `+` | Addition | `12 + 3` | 15 |
| `-` | Subtraction | `12 - 3` | 9 |
| `*` | Multiplication | `12 * 5` | 60 |
| `/` | Division | `12 / 5` | 2 |
| `%` | Remainder | `12 % 5` | 2 |

Parentheses group a calculation. Multiplication, division, and remainder bind more tightly than addition and subtraction. Operators of those arithmetic groups associate left to right. Trace intermediate results instead of treating a long expression as one unexplained formula.


In [1]:
int quantity = 12;
int extraCases = 3;
int unitsPerCase = 10;
Console.WriteLine(quantity + extraCases * unitsPerCase);
Console.WriteLine((quantity + extraCases) * unitsPerCase);


42
150


The first result is 42; the second is 150. The second matches the requirement “add the extra cases, then count all units.” Readability matters: named intermediate results such as `revisedQuantity` can make the intended grouping easier to review.


```{index} integer division; remainder; packing boxes
```

(division-and-remainders)=

### Integer Division and Remainders



With integer operands, division truncates toward zero. For a nonnegative order of 12 cases and five cases per box, the quotient is two full boxes and the remainder is two cases. For positive box capacity, `quantity = fullBoxes * casesPerBox + leftoverCases` reconstructs the original quantity.

Do not call the quotient the total number of boxes needed when leftovers need another box. The remainder is not a fractional box.


In [1]:
int quantity = 12;
int casesPerBox = 5;
int fullBoxes = quantity / casesPerBox;
int leftoverCases = quantity % casesPerBox;
Console.WriteLine(fullBoxes);
Console.WriteLine(leftoverCases);
Console.WriteLine(fullBoxes * casesPerBox + leftoverCases);
Console.WriteLine(-12 / 5);
Console.WriteLine(-12 % 5);


2
2
12
-2
-2


The negative examples give -2 and -2. For integer remainder in C#, a nonzero result has the dividend's sign; it is not always a nonnegative mathematical modulo. Division by integer zero throws an exception. The packing model requires nonnegative quantity and a strictly positive box capacity.[^1]


```{index} division type; conversion timing
```

### Convert an Operand Before Division



Assigning an integer division result to `double` cannot recover the fraction already discarded. Convert an operand before the operation if the requirement asks for a fractional ratio.


In [1]:
using System.Globalization;
CultureInfo.CurrentCulture = CultureInfo.InvariantCulture;

int quantity = 12;
int casesPerBox = 5;
double truncated = quantity / casesPerBox;
double fractional = (double)quantity / casesPerBox;
Console.WriteLine(truncated);
Console.WriteLine(fractional);
Console.WriteLine((6.0 / 3.0).GetType());


2
2.4
System.Double


The values are two and 2.4. `6.0 / 3.0` still has type `double` even though printing it without a format may show `2`. The displayed digits do not determine the value's type.


In [ ]:
// ### Exercise: Check a Packing Calculation
// For 23 cases and capacity five, calculate fullBoxes and leftoverCases. Print both and the reconstructed quantity. Expected: 4, 3, 23.
// Your code starts here.

// Your code ends here.


In [1]:
// ### Solution
int quantity = 23;
int capacity = 5;
int fullBoxes = quantity / capacity;
int leftoverCases = quantity % capacity;
Console.WriteLine(fullBoxes);
Console.WriteLine(leftoverCases);
Console.WriteLine(fullBoxes * capacity + leftoverCases);


4
3
23


```{index} assignment operators; compound assignment; increment
```

(assignment-operators)=

## Assignment Operators Change Variables



`quantity += 3` adds three and assigns the new value to `quantity`. For the simple `int` variables here, it has the same effect as `quantity = quantity + 3`. `-=`, `*=`, `/=`, and `%=` similarly combine an operation with assignment. More complex operands and conversions need additional care, so the shorthand is not an unrestricted text replacement rule.


In [1]:
int quantity = 12;
quantity += 3;
quantity -= 2;
Console.WriteLine(quantity);
quantity++;
Console.WriteLine(quantity);


13
14


`++` increases a numeric variable by one; `--` decreases it by one. When each is its own statement, prefix and postfix forms produce the same updated variable. Inside a larger expression, postfix supplies the old value and prefix supplies the new value. Prefer separate statements while learning so state changes are visible.


In [1]:
int quantity = 12;
int previous = quantity++;
Console.WriteLine(previous);
Console.WriteLine(quantity);
int updated = ++quantity;
Console.WriteLine(updated);
Console.WriteLine(quantity);


12
13
14
14


```{index} comparison operators; equality
```

## Comparisons Produce Boolean Values



`==` tests equality; `!=` tests inequality. `<`, `<=`, `>`, and `>=` compare order. These expressions produce `bool` values without changing the quantities being compared. `=` assigns a value; it does not ask a question.


In [1]:
int quantity = 12;
int minimum = 10;
bool freeShipping = quantity >= minimum;
Console.WriteLine(freeShipping);
Console.WriteLine(quantity == minimum);
Console.WriteLine(quantity != minimum);


True
False
True


```{index} logical operators; short-circuit evaluation
```

(logical-operators)=

### Combining Conditions and Short-Circuit Evaluation



`&&` means both Boolean operands must be true; `||` means at least one must be true; `!` negates a Boolean. A free-shipping rule might require a large enough order **and** an active customer account.

| Left | Right | Left && Right | Left or Right |
| --- | --- | --- | --- |
| false | false | false | false |
| false | true | false | true |
| true | false | false | true |
| true | true | true | true |

The right operand of `&&` is skipped when the left is false. The right operand of `||` is skipped when the left is true. This **short-circuit evaluation** can guard an operation that would otherwise fail.


In [1]:
int quantity = 12;
bool activeAccount = true;
bool qualifies = quantity >= 10 && activeAccount;
Console.WriteLine(qualifies);
int denominator = 0;
bool enoughPerGroup = denominator != 0 && quantity / denominator >= 2;
Console.WriteLine(enoughPerGroup);


True
False


The second result is `False`; division is skipped because `denominator != 0` is false. Reversing the operands would attempt the division first and fail. This is a reason to understand evaluation order, not a substitute for designing complete input validation. Branching with `if` belongs in Chapter 4.

A numeric range test is `quantity >= 0 && quantity <= 100`. Writing `0 <= quantity <= 100` does not express a chained comparison in C#.


```{index} operator precedence; left associativity
```

(precedence)=

## Precedence and a Reviewable Calculation



For the operators used in the core examples, higher rows below bind more tightly. Use parentheses to communicate intent when mixing groups. Precedence determines grouping; short-circuit rules determine whether a grouped Boolean operand is evaluated.

| Group | Operators |
| --- | --- |
| Grouping / member access | parentheses, `.` |
| Unary | `!`, unary `-`, casts |
| Multiplicative | `*`, `/`, `%` |
| Additive | `+`, `-` |
| Ordering | `<`, `<=`, `>`, `>=` |
| Equality | `==`, `!=` |
| Conditional AND | `&&` |
| Conditional OR | `\|\|` |
| Assignment | `=`, `+=`, `-=` and related forms |

This is a working subset, not the full language table. For an invoice, use named stages: subtotal, discount, discounted amount, rounded tax, and total. That makes the business order explicit: discount first, then calculate sample tax on the discounted amount.


In [1]:
using System.Globalization;
CultureInfo.CurrentCulture = CultureInfo.InvariantCulture;

int quantity = 12;
decimal unitPrice = 38.50m;
decimal discountRate = 0.05m;
decimal taxRate = 0.07m;
decimal subtotal = quantity * unitPrice;
decimal discount = Math.Round(subtotal * discountRate, 2, MidpointRounding.AwayFromZero);
decimal discounted = subtotal - discount;
decimal tax = Math.Round(discounted * taxRate, 2, MidpointRounding.AwayFromZero);
decimal total = discounted + tax;
Console.WriteLine(subtotal);
Console.WriteLine(discount);
Console.WriteLine(tax);
Console.WriteLine(total);


462.00
23.10
30.72
469.62


```{index} bitwise operators; bit shifts
```

### Optional: Operations on Bits



Integer `&`, `|`, and `^` combine corresponding bits using AND, OR, and exclusive OR. They differ from conditional Boolean `&&` and `||`. Here, five is binary `0101` and three is `0011`.


In [1]:
int left = 5;
int right = 3;
Console.WriteLine(left & right);
Console.WriteLine(left | right);
Console.WriteLine(left ^ right);
Console.WriteLine(left << 1);
Console.WriteLine(left >> 1);


1
7
6
10
2


The results are one, seven, six, 10, and two. For these small positive values, shifting left by one doubles the value and shifting right by one discards its low bit. Do not generalize shifts to every multiplication or division: overflow, negative values, and shift-count rules matter. You do not need bit manipulation for the invoice lab.

**Exit check:** Explain why integer division, assignment, and equality answer different questions. Then justify both the parentheses and the rounding stages in the invoice.


```{rubric} Footnotes
```

[^1]: The [C# arithmetic operators](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/arithmetic-operators) reference specifies integer division and remainder behavior. The [operators reference](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/) provides the complete precedence table.
